### PR TITLE
Implements Bayou Vermilion Railroad

### LINKED ISSUE #657: O4B - Bayou Vermilion Railroad

### PR DESCRIPTION
**Pull request by stopfstedt**
fixes #657

### CODE REVIEW

**Comment by stopfstedt** [comment]
@mmeldo this seems to work now, for the most part. 

the remaining issue is that the "untilEndOfPhase" effect doesn't stick once the _Mystical_ cards is moved off of the location that got nerfed. 

Example: 
1. My dude with Evanor moves to my opponent's Railway Station
2. I use the Bayou Vermilion Railroad's noon ability with Evanor to tank the Railway Station's CP and production.
3. My dude moves elsewhere - the railway station now regains its original CP and prod.

I need to make the initial location (Railway Station) stick somehow. Any suggestions?

**Comment by mmeldo** [comment]
> @mmeldo this seems to work now, for the most part.
> 
> the remaining issue is that the "untilEndOfPhase" effect doesn't stick once the _Mystical_ cards is moved off of the location that got nerfed.
> 
> Example:
> 
> 1. My dude with Evanor moves to my opponent's Railway Station
> 2. I use the Bayou Vermilion Railroad's noon ability with Evanor to tank the Railway Station's CP and production.
> 3. My dude moves elsewhere - the railway station now regains its original CP and prod.
> 
> I need to make the initial location (Railway Station) stick somehow. Any suggestions?

The effects can be state dependent or not. State dependent effects get reapplied after every event, which sometime (as in your case) is something you do not want. To make a state dependent effect, there are 2 ways:
- fixed: `isStateDependent: true` property defined for the effect (e.g. any dynamic stat, for example `dynamicBullets`)
- by specifying `condition` along with `match` being a function

What you have is the second one which means the effect is state dependent. If you want to have it applied only once and be done, remove the `condition` and pass a card or array of cards in the `match` instead of the function. So in our case, it should be enough to have `match: context.target.locationCard`

**Comment by mmeldo** [comment]
I did a small refactor regarding the location functions, I hope you do not mind. The problem here is that using the `getGameLocation()` is dangerous as it can return `null` and cause crashes. Thus it is better to put it into the helper functions with proper checks.

File: server/game/cards/13-O4B/BayouVermilionRailroad.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            title: 'Bayou Vermilion Railroad',
+            playType: 'noon',
+            cost: ability.costs.bootSelf(),        
+            target: {
+                activePromptTitle: 'Select a Mystical card',
+                cardCondition: { 
+                    location: 'play area',
+                    condition: card => card.hasKeyword('mystical') && card.isAtDeed() && !card.booted && card.canBeBooted()
+                }
+            },
```
Comment (mmeldo) [inline server/game/cards/13-O4B/BayouVermilionRailroad.js:20]: It is cleaner to use `gameAction` as that is used in the engine to check if the action is allowed and some other validation.
```suggestion
            target: {
                activePromptTitle: 'Select a Mystical card',
                cardCondition: { 
                    location: 'play area',
                    condition: card => card.hasKeyword('mystical') && card.isAtDeed()
                },
                gameAction: 'boot'
            },
```

File: server/game/cards/13-O4B/BayouVermilionRailroad.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            handler: context => {
+                this.game.resolveGameAction(GameActions.bootCard({ card: context.target }), context);
+                context.player.modifyGhostRock(1);
+                this.untilEndOfPhase(context.ability, ability => ({
+                    match: context.target.locationCard,
+                    effect: [
+                        ability.effects.setProduction(0),
+                        ability.effects.setControl(0)
+                    ]
+                }), 'upkeep');
```
Comment (mmeldo) [inline server/game/cards/13-O4B/BayouVermilionRailroad.js:30]: > you may reduce the deed's control points and production to 0 until after the next Upkeep phase

It says you may reduce, so we should put a prompt to ask player if he wants to reduce.

**Review by mmeldo — CHANGES_REQUESTED** [review]

File: server/game/cards/13-O4B/BayouVermilionRailroad.js (hunk 2 above)
Comment (stopfstedt) [reply server/game/cards/13-O4B/BayouVermilionRailroad.js:31]: good catch, i'll wrap it into a prompt tonight.

Comment (stopfstedt) [reply server/game/cards/13-O4B/BayouVermilionRailroad.js]: I added the prompt and some game messages.

**Review by mmeldo — APPROVED** [review]

**Pull request merged**